src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                        });
+                    }
+                };
+
+                // Check guard is boolean
+                let guard_type = typeck_expr(
+                    &case.guard,
+                    &TypeCtx {
+                        env: case_env.clone(),
+                        funcs: ctx.funcs,
+                    },
+                )?;
+                if guard_type != Type::BoolT {
+                    return Err(TypeError::TypeMismatch {
+                        expected: Type::BoolT,
+                        actual: guard_type,
+                    });
+                }
+
+                // Check result type
+                let case_result_type = typeck_expr(
+                    &case.result,
+                    &TypeCtx {
+                        env: case_env,
+                        funcs: ctx.funcs,
+                    },
+                )?;
+
+                // Ensure consistency with previous cases
+                if let Some(ref t) = result_type {
+                    if *t != case_result_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: t.clone(),
+                            actual: case_result_type,
+                        });
+                    }
+                } else {
+                    result_type = Some(case_result_type);
+                }
+            }
+
+            // Return the type of the result
+            result_type.ok_or(TypeError::MismatchPatternScrutinee {
+                expected: scrutinee_type,
+                actual: cases[0].pattern.clone(),
+            })
+        }
+    }
+}
+
+/// Typecheck a statement in a given environment
+pub fn typeck_stmt(stmt: &Stmt, ctx: &TypeCtx) -> Result<(), TypeError> {
+    // Either an asignment statemnt (x = 5)
+    // Or an if statment (if condition - true block - or  - false block)
+    // need to be able to assign a var or an expression
+
+    match stmt {
+        Stmt::Assign(id, expr) => {
+            // Get the type of the variable from the environment
+            let var_type = ctx
+                .env
+                .get(id)
+                .cloned()
+                .ok_or(TypeError::UndefinedVariable(*id))?;
+
+            // Type check the expression
+            let expr_type = typeck_expr(expr, ctx)?;
+
+            // Check that the expression's type matches the variable's type
+            if var_type != expr_type {
+                return Err(TypeError::TypeMismatch {
+                    expected: var_type,
+                    actual: expr_type,
+                });
+            }
+            Ok(())
+        }
+        // If statement: check that the condition is a boolean and type check both branches
+        Stmt::If {
+            cond,
+            true_branch,
+            false_branch,
+        } => {
+            // Type check the condition
+            let cond_type = typeck_expr(cond, ctx)?;
```

Comment:
note: once you remove the whole `expected` stuff, you can get rid of all these short-lived variables and move the small calls to their uses.